src/Antiforgery/src/AntiforgeryMiddleware.cs

```diff
@@ -0,0 +1,130 @@
+    {
+        _antiforgery = antiforgery;
+        _next = next;
+        _logger = logger;
+    }
+
+    public Task Invoke(HttpContext context)
+    {
+        var endpoint = context.GetEndpoint();
+        if (endpoint is null)
+        {
+            return _next(context);
+        }
+
+        var antiforgeryMetadata = endpoint.Metadata.GetMetadata<IAntiforgeryMetadata>();
+        if (antiforgeryMetadata is null)
+        {
+            Log.NoAntiforgeryMetadataFound(_logger);
+            return _next(context);
+        }
+        
+        if (antiforgeryMetadata is not IValidateAntiforgeryMetadata validateAntiforgeryMetadata)
+        {
+            Log.IgnoreAntiforgeryMetadataFound(_logger);
+            return _next(context);
+        }
+
+        if (_antiforgery is DefaultAntiforgery defaultAntiforgery)
+        {
+            var valueTask = defaultAntiforgery.TryValidateAsync(context, validateAntiforgeryMetadata.ValidateIdempotentRequests);
+            if (valueTask.IsCompletedSuccessfully)
+            {
+                var (success, message) = valueTask.GetAwaiter().GetResult();
+                if (success)
+                {
+                    Log.AntiforgeryValidationSucceeded(_logger);
+                    return _next(context);
+                }
+                else
+                {
+                    Log.AntiforgeryValidationFailed(_logger, message);
+                    return WriteAntiforgeryInvalidResponseAsync(context, message);
+                }
+            }
+
+            return TryValidateAsyncAwaited(context, valueTask);
+        }
+        else
+        {
+            return ValidateNonDefaultAntiforgery(context);
+        }
+    }
+
+    private async Task TryValidateAsyncAwaited(HttpContext context, ValueTask<(bool success, string? message)> tryValidateTask)
+    {
+        var (success, message) = await tryValidateTask;
+        if (success)
+        {
+            Log.AntiforgeryValidationSucceeded(_logger);
+            await _next(context);
+        }
+        else
+        {
+            Log.AntiforgeryValidationFailed(_logger, message);
+            await context.Response.WriteAsJsonAsync(new ProblemDetails
+            {
+                Status = StatusCodes.Status400BadRequest,
+                Title = "Antiforgery validation failed",
+                Detail = message,
+            });
+        }
+    }
+
+    private async Task ValidateNonDefaultAntiforgery(HttpContext context)
+    {
+        if (await _antiforgery.IsRequestValidAsync(context))
+        {
+            Log.AntiforgeryValidationSucceeded(_logger);
+            await _next(context);
+        }
+        else
+        { 
+            Log.AntiforgeryValidationFailed(_logger, message: null);
+            await WriteAntiforgeryInvalidResponseAsync(context, message: null);
+        }
+    }
+
+    private static Task WriteAntiforgeryInvalidResponseAsync(HttpContext context, string? message)
```

Comment:
Why is this JSON?

Reply:
Seems wrong to be forcibly returning JSON here. Why wouldn't we just let it throw or throw `BadHttpRequestException` or something else to enable the app to handle how this responds to the client?

Reply:
Or make it an event.